pkg/fqdn/cache.go

```diff
@@ -817,12 +836,27 @@ func (zombies *DNSZombieMappings) Upsert(expiryTime time.Time, addr netip.Addr,
 }
 
 // isConnectionAlive returns true if 'zombie' is considered alive.
-// Zombie is considered dead if both of these conditions apply:
+// Zombie is considered dead if all of these conditions apply:
 // 1. CT GC has run after the DNS Expiry time and grace period (lastCTGCUpdate > DeletePendingAt + GracePeriod), and
-// 2. The CG GC run did not mark the Zombie alive (lastCTGCUpdate > AliveAt)
+// 2. The CT GC run did not mark the Zombie alive (lastCTGCUpdate > AliveAt)
+// 3. CT GC has run at least 2 times since Zombie was entered
 // otherwise the Zombie is alive.
+//
+// We wait for 2 complete GC runs, because this entry may have been added in the middle of a GC run,
+// in which case it may not have been marked alive. We need to wait for GC to finish at least 2 times
+// before we can safely consider it dead.
 func (zombies *DNSZombieMappings) isConnectionAlive(zombie *DNSZombieMapping) bool {
-	return !(zombies.lastCTGCUpdate.After(zombie.DeletePendingAt) && zombies.lastCTGCUpdate.After(zombie.AliveAt))
+	if !zombies.lastCTGCUpdate.After(zombie.DeletePendingAt.Add(option.Config.ToFQDNsIdleConnectionGracePeriod)) {
```

Comment:
Interesting that the grace period was missing from the computation while it was already documented in the comment above.

Reply:
there are a lot of interesting things about FQDN GC.